app/services/FilterService.scala

```diff
@@ -0,0 +1,139 @@
+      if (filtered.nonEmpty && filtered.get.nonEmpty && numberOfFiltered >= 0) {
+        logger.info(s"[FilterService][filterEstimatedLPP1DuringPeriodOfFamiliarisation] Filtering for [$callingClass][$function] -" +
+          s" Filtered ${numberOfFiltered} LPP1(s) from payload for VRN: $vrn")
+        penaltiesDetails.copy(latePaymentPenalty = Some(LatePaymentPenalty(filtered)))
+      } else {
+        logger.info(s"[FilterService][filterEstimatedLPP1DuringPeriodOfFamiliarisation] Filtering for [$callingClass][$function] - Filtered all LPP1s from payload for VRN: $vrn")
+        penaltiesDetails.copy(latePaymentPenalty = None)
+      }
+    } else {
+      logger.info(s"[FilterService][filterEstimatedLPP1DuringPeriodOfFamiliarisation] Filtering for [$callingClass][$function] - No LPPs to filter for VRN: $vrn")
+      penaltiesDetails
+    }
+  }
+
+  private def filterEstimatedLPP1(penaltiesDetails: GetPenaltyDetails): Option[Seq[LPPDetails]] = {
+    penaltiesDetails.latePaymentPenalty.flatMap(
+      _.details.map(latePaymentPenalties => latePaymentPenalties.filterNot(lpp => {
+        lpp.penaltyCategory.equals(LPPPenaltyCategoryEnum.FirstPenalty) &&
+          lpp.penaltyStatus.equals(LPPPenaltyStatusEnum.Accruing) &&
+          appConfig.withinLPP1FilterWindow(lpp.principalChargeDueDate)
+      })
+      )
+    )
+  }
+
+  def filterPenaltiesWith9xAppealStatus(penaltiesDetails: GetPenaltyDetails)(implicit callingClass: String, function: String, vrn: String): GetPenaltyDetails = {
+    val filteredLSPs: Option[Seq[LSPDetails]] = if (penaltiesDetails.lateSubmissionPenalty.nonEmpty) filterLSPWith9xAppealStatus(penaltiesDetails) else None
+    val noOfFilteredLSPs:Int = numberOfFilteredLSPs(filteredLSPs, penaltiesDetails)
+    val filteredLPPs: Option[Seq[LPPDetails]] = if (penaltiesDetails.latePaymentPenalty.nonEmpty) filterLPPWith9xAppealStatus(penaltiesDetails) else None
+    val noOfFilteredLPPs: Int = numberOfFilteredLPPs(filteredLPPs, penaltiesDetails)
+
+    (noOfFilteredLSPs, noOfFilteredLPPs) match {
+      case (x , 0) if x > 0 => penaltiesDetails.copy(lateSubmissionPenalty = prepareLateSubmissionPenaltiesAfterFilter(penaltiesDetails, filteredLSPs, noOfFilteredLSPs, vrn, callingClass, function))
+      case (0, x) if x > 0 =>
+        penaltiesDetails.copy(latePaymentPenalty = prepareLatePaymentPenaltiesAfterFilter(filteredLPPs, noOfFilteredLPPs, vrn, callingClass, function))
+      case (x, y) if x > 0 && y > 0 =>
+        val newLasteSubmissions: Option[LateSubmissionPenalty] = prepareLateSubmissionPenaltiesAfterFilter(penaltiesDetails, filteredLSPs, noOfFilteredLSPs, vrn, callingClass, function)
+        val newLastePaymentPenalties: Option[LatePaymentPenalty] = prepareLatePaymentPenaltiesAfterFilter(filteredLPPs, noOfFilteredLPPs, vrn, callingClass, function)
+        penaltiesDetails.copy(lateSubmissionPenalty = newLasteSubmissions, latePaymentPenalty = newLastePaymentPenalties)
+      case _ =>
```

Comment:
This code is quite convoluted, is there not a generic way we can return these LSPs and LPPs. I'm conscious of the effect this will have on performance